### ISSUE TITLE
sema: pending scope flags land on a while loop's continue expression instead of its body

### ISSUE THREAD

**Issue by DonIsaac** [DonIsaac/zlint#400]
## Summary

`visitWhile` visits a `while` loop's continue expression *before* its body:

https://github.com/DonIsaac/zlint/blob/main/src/Semantic/Builder.zig#L884-L885

```zig
try self.visitOptional(ast.cont_expr);
try self.visit(ast.then_expr);
```

`_next_block_scope_flags` is a one-shot channel: `visitBlock` merges it into the
next block scope it creates and then clears it (`Builder.zig:525-526`). So when a
pending flag is set by an enclosing construct — `visitCatch:1066` sets
`s_catch` before visiting its fallback expression — and that `while`'s
`cont_expr` happens to be a block (`: ({ ... })`), the continue block consumes
the flag and the loop body is left without it.

The flag lands on the wrong scope in both directions: the continue expression
gains a flag it should never have, and the body loses one it should have.

## How it impacts users

No shipped lint rule reads the `s_catch` *scope* flag today (`grep -rn s_catch src/`
finds only `Builder.zig`, `Scope.zig`, `Symbol.zig`, and one `s_catch_param`
assertion in `symbol_decl_test.zig`), so there is **no user-visible
misdiagnosis right now**. This is latent-correctness, not a live false
positive/negative.

It matters because it is a trap for the next rule that needs "am I inside a
`catch` block?". Such a rule would silently mis-answer for exactly one shape —
a `catch` whose fallback is a `while` with a block continue expression —
producing a false negative on the body and a false positive on the continue
expression. That is an awkward bug to discover from a rule author's seat,
since nothing about the rule would be wrong.

It also makes the scope tree an inaccurate model of the source, which is the
one thing `src/Semantic/` exists to provide.

## Reproduction steps

`src/repro.zig` (must live under the repo so the CLI walker picks it up):

```zig
fn bar() anyerror!u32 {
    return 1;
}

fn foo() u32 {
    var i: u32 = 0;
    return bar() catch while (i < 10) : ({
        i += 1;
    }) {
        const inner = i;
        _ = inner;
    } else 0;
}
```

```sh
zig build install
./zig-out/bin/zlint --print-ast src/repro.zig
```

Observed scope tree — scope 7 is the continue block `({ i += 1; })`, scope 8 is
the loop body holding `inner`:

```json
{ "id": 5, "flags": ["catch"],                    // visitCatch's payload scope
  "children": [
    { "id": 6, "flags": [],                       // while payload scope
      "children": [
        { "id": 7, "flags": ["block", "catch"] },  // ← cont_expr, wrongly tagged
        { "id": 8, "flags": ["block"],             // ← body, missing `catch`
          "bindings": { "inner": 4 } }
      ] } ] }
```

Expected: `"catch"` on scope 8, absent from scope 7.

## How to test

A unit test at the bottom of `src/Semantic/Builder.zig` is the right home —
`Builder.zig:2005` already asserts `block_scope.flags.s_block` and is the
precedent for reaching into the scope tree directly.

Assert both halves, since a fix that only moves the flag forward could still
leave it on the continue block:

1. Build the semantic model for the source above.
2. Find the scope that binds `inner` (the loop body) — assert `flags.s_catch`.
3. Find its sibling scope under the same parent (the continue block) — assert
   `!flags.s_catch`.

The existing `whileWithExpr` fixture in
`test/fixtures/simple/pass/loops_while.zig` already has a block continue
expression, so wrapping its `while` in a `catch` there would give e2e snapshot
coverage too — though that would renumber scope IDs in
`loops_while.zig.snap`, so the unit test is the cheaper guard.

## How to fix

**Do not** simply swap the two `visit` calls. Visiting `then_expr` first would
match Zig's evaluation order, but it renumbers every scope ID downstream of any
`while` with a `cont_expr`, churning `loops_while.zig.snap` and
`fibonacci.zig.snap` for no semantic gain.

Instead, withhold the pending flags across the `cont_expr` visit and restore
them for the body — the continue expression is not the loop body and should
never claim the body's pending flags:

```zig
        {
            const pending = self._next_block_scope_flags;
            self._next_block_scope_flags = .{};
            defer self._next_block_scope_flags = pending;
            try self.visitOptional(ast.cont_expr);
        }
        try self.visit(ast.then_expr);
```

Verified locally: this puts `["block", "catch"]` on the body and `["block"]` on
the continue block, and `just e2e` produces **zero** snapshot churn across all
20 simple fixtures and the four ecosystem suites.

While fixing, check whether `visitFor` needs the same treatment — it has no
`cont_expr`, so it looks safe, but `visitWhile`'s then-payload scope and
`visitIf`'s branches share this one-shot-channel pattern and are worth an audit
pass.

<sub>Found while reviewing #376 (payload captures and `extern fn` names binding
wrong); a `TODO` marking this spot is included in that PR.</sub>
